### ISSUE TITLE
Service w/ sessionAffinity clientIp not respecting x-forwarded-for header

### ISSUE THREAD

**Issue by tjmehta** [kubernetes/kubernetes#36415]
I assume this is a bug, as most load balancers respect the `x-forwarded-for` header.
But the documentation makes no mention of weather `sessionAffinity` `clientIp` will "stick" requests by `x-forwarded-*` headers.

**Kubernetes version**
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"clean", BuildDate:"2016-10-10T18:19:49Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:32:42Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
Google Cloud Platform - GKE
Default google cluster image

**What happened**:
Requests are "sticking" based on socket remote-address

**What you expected to happen**:
Requests should stick based on `x-forwarded-for` header

**How to reproduce it** (as minimally and precisely as possible):
Create a service on GKE w/ `type: LoadBalancer` and `sessionAffinity: clientIp`
Create 2 services that log incoming requests
`curl -v -H x-forwarded-for=1.1.1.1  <ip>` // should hit a pod
`curl -v -H x-forwarded-for=1.1.1.1  <ip>` // should hit the same pod

**Comment by ApsOps** [comment]
I think I'm hitting this as well.

What header does kube-proxy use anyway to keep session affinity?

**Comment by solsson** [comment]
@ApsOps what do yo mean? That `sessionAffinity`=`ClientIP` can't make use of the X-Forwarded-For header? In #2867, then what do you mean with "pass the real IP to your service"? And from where is the quote above?

The reason I'm asking is that I have a service that I'm quite sure had stickyness based on `X-Forwarded-For` in k8s 1.4.x but I can no longer conclude that it works, and I can't find a reference to sessionAffinity + headers in docs. The source in both [1.5](https://github.com/kubernetes/kubernetes/blob/v1.5.3/pkg/proxy/userspace/roundrobin.go#L144) and [1.4](https://github.com/kubernetes/kubernetes/blob/v1.4.4/pkg/proxy/userspace/roundrobin.go#L144) says "Don't shadow IP".

Is the strategy in your blog post still valid?

**Comment by ApsOps** [comment]
@solsson Turns out some of the info I had given was incorrect. Sorry about that.

You can check the full discussion here - https://groups.google.com/d/topic/kubernetes-sig-network/NUH7co8NDAQ/discussion

**Comment by solsson** [comment]
Based on that thread, is your conclusion like mine: only the actual remote IP of the TCP connection can be used with sessionAffinity=ClientIP?

For socket.io etc without Ingress (nginx-sticky-module-ng + bypasses services) this basically means you'd need a service that is `type`=`LoadBalancer` AND  `"service.beta.kubernetes.io/external-traffic": "OnlyLocal"` AND `sessionAffinity`=`ClientIP`, correct?

To support "proxy protocol" LB, someone would have to invent a way to get remote IP on the TCP level? And `X-Forwarded-For` or `X-Real-Ip` are actually out of the question?

There are some mentions of these two headers in the source, like "Looks at X-Forwarded-For header, X-Real-Ip header and request.RemoteAddr in that order.", but is that only for apiserver?

**Comment by ApsOps** [comment]
@solsson With the recent info I got, I understand that it's not possible to look at headers currently.

The reason why my setup works (the one I specified in the blog post) is that:

Since `sessionAffinity` is enabled at both intermediary nginx and socket.io container, the connection from an ELB server comes to the same nginx pod every time, and connection from an nginx pod goes to the same socket.io pod every time.

I'd try and investigate more to get a clear picture.
